packages/insured-bridge-relayer/src/Relayer.ts

```diff
@@ -56,73 +64,165 @@ export class Relayer {
+            await this.slowRelay(relayableDeposit.deposit, realizedLpFeePct);
             break;
 
-          case RelayType.Fast:
-            this.logger.info({
-              at: "Relayer",
-              message: "Fast relaying deposit 🏃‍♂️",
-              l1Token,
+          case RelayType.SpeedUp:
+            this.logger.debug({
+              at: "InsuredBridgeRelayer#Relayer",
+              message: "Speeding up existing relayed deposit",
+              realizedLpFeePct,
               relayableDeposit,
-              botL1TokenBalance,
             });
-            await this.fastRelay(relayableDeposit);
+            await this.speedUpRelay(relayableDeposit.deposit);
+            break;
+          case RelayType.Instant:
+            this.logger.debug({
+              at: "InsuredBridgeRelayer#Relayer",
+              message: "Instant relaying deposit",
+              realizedLpFeePct,
+              relayableDeposit,
+            });
+            await this.instantRelay(relayableDeposit.deposit);
             break;
         }
       }
     }
   }
 
-  private shouldRelay(deposit: Deposit, botTokenBalance: string): RelayType {
-    // shouldRelay(deposit) => RelayType { No, Slow, Fast }
-    // Checks bot balance.
-    // If bot balance is > transfer size, simulate the fast relay to determine the cost.
-    // If bot balance is > bond, simulate the slow relay to determine the cost.
-    // If the max profit across the two (or all that succeeded) is negative or 0, return No.
-    // Else return the one with the most profit.
-
-    // TODO: Add logic to decide if a relay action should happen
-    console.log(deposit, botTokenBalance);
-    return RelayType.Slow;
+  private async shouldRelay(deposit: Deposit, relayAbility: RelayAbility, realizedLpFeePct: BN): Promise<RelayType> {
+    const [l1TokenBalance, proposerBondPct] = await Promise.all([
+      getTokenBalance(this.l1Client.l1Web3, deposit.l1Token, this.account),
+      this.l1Client.getProposerBondPct(),
+    ]);
+    const relayTokenRequirement = this.getRelayTokenRequirement(deposit, proposerBondPct, realizedLpFeePct);
+
+    // There are three different kinds of profits that the bot can produce:
+    let slowProfit = toBN("0");
+    let speedUpProfit = toBN("0");
+    let instantProfit = toBN("0");
+    // Based on the bots token balance, and the relay state we can compute the profitability of each action.
+
+    // a) Balance is large enough to do a slow relay. No relay action has happened on L1 yet for this deposit.
+    if (l1TokenBalance.gte(relayTokenRequirement.slow) && relayAbility === RelayAbility.Any)
+      slowProfit = toBN(deposit.amount).mul(toBN(deposit.slowRelayFeePct)).div(fixedPointAdjustment);
+
+    // b) Balance is large enough to instant relay. Deposit is in any state except finalized (i.e can be slow relayed
+    // and sped up or only sped up.)
+    if (l1TokenBalance.gte(relayTokenRequirement.instant) && relayAbility !== RelayAbility.None)
```

Comment:
This is a case where renamed enums like I mention [here](https://github.com/UMAprotocol/protocol/pull/3359/files#r707783410) would improve readability.

Reply:
for sure. it looks better now.